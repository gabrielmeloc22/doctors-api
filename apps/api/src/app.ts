import bodyParser from "body-parser"
import cors from "cors"
import express, { type Express } from "express"
import morgan from "morgan"

export const createApp = (): Express => {
    const app = express()

    app.disable("x-powered-by")
        .use(morgan("dev"))
        .use(bodyParser.urlencoded({ extended: true }))
        .use(bodyParser.json())
        .use(cors())

    return app
}
