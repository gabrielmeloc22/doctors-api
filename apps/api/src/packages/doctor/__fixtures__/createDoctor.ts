import { db } from "../../../database/db";
import { DeepPartial } from "../../../utils/deepPartial";
import { doctorTable, IDoctor } from "../doctorTable";

export const createDoctor = async (
    args?: DeepPartial<IDoctor>
): Promise<IDoctor> => {
    const doctorCount = await db.$count(doctorTable);

    const [doctor] = await db
        .insert(doctorTable)
        .values({
            email: `doc${doctorCount}@mail.com`,
            firstName: "john",
            lastName: `doe ${doctorCount}`,
            username: `john#${doctorCount}`,
            ...args,
        })
        .returning();

    return doctor as IDoctor;
};
