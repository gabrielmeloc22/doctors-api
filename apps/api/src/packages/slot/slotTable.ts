import * as t from "drizzle-orm/pg-core";
import { idField, timestampField } from "../database/schemaDefaults";
import { doctorTable } from "../doctor/doctorTable";

export const slotTable = t.pgTable("slot", {
    doctorId: t.uuid().references(() => doctorTable.id),
    startTime: t.date().notNull(),
    endTime: t.date().notNull(),
    available: t.boolean(),
    ...timestampField,
    ...idField,
});
