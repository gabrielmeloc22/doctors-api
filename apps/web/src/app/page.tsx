"use client"

import { ReactNode } from "react"

export default function Web(): ReactNode {
    return <h1>batata</h1>
}
