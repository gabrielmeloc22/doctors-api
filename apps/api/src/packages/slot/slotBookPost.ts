import { getDay, getHours } from "date-fns";
import { and, eq } from "drizzle-orm";
import { RequestHandler } from "express";
import { z } from "zod";
import { db } from "../../database/db";
import { Result } from "../../utils/result";
import { validateRequest } from "../../utils/validateRequest";
import { bookingTable } from "../booking/bookingTable";
import { SLOT_REPEAT_TYPE_ENUM, slotTable } from "../slot/slotTable";
import { getTimeSlot } from "./lib/getTimeSlot";
import { getTimeSlots } from "./lib/getTimeSlots";

const slotBookPostParams = z.object({
    id: z.string().uuid(),
});

type SlotBookPostParams = z.infer<typeof slotBookPostParams>;

const slotBookPostBodySchema = z.object({
    reason: z.string(),
    patient_id: z.string(),
    start_time: z.string().datetime(),
});

type SlotBookPostBody = z.infer<typeof slotBookPostBodySchema>;

type SlotBookPostResult = Result<{
    booking: {
        id: string;
        slot_id: string;
        patient_id: string;
        reason: string;
        time: string;
    };
}>;

const slotBookPostHandler: RequestHandler<
    SlotBookPostParams,
    SlotBookPostResult,
    SlotBookPostBody
> = async (req, res) => {
    const [slot] = await db
        .select()
        .from(slotTable)
        .where(eq(slotTable.id, req.params.id));

    if (!slot) {
        res.status(404);
        res.json({ success: false, error: "Slot not found" });
        return;
    }

    // TODO: create recurrence rule validator

    const selectedStart = new Date(req.body.start_time);

    if (selectedStart < slot.startTime) {
        res.status(400);
        res.json({ success: false, error: "Booking time not available" });
        return;
    }

    if (slot.repeatEnd && selectedStart > new Date(slot.repeatEnd)) {
        res.status(400);
        res.json({ success: false, error: "Booking time not available" });
        return;
    }

    if (
        slot.repeatType === SLOT_REPEAT_TYPE_ENUM.WEEKLY &&
        !slot.repeatWeekdays?.includes(getDay(selectedStart))
    ) {
        res.status(400);
        res.json({ success: false, error: "Booking time not available" });
        return;
    }

    if (getHours(selectedStart) > getHours(slot.endTime)) {
        res.status(400);
        res.json({ success: false, error: "Booking time not available" });
        return;
    }

    const timeSlots = getTimeSlots(slot);

    const selectedTimeSlot = getTimeSlot(slot, selectedStart);

    // check if booking time is within time slots of the day
    if (timeSlots[selectedTimeSlot] === undefined) {
        res.status(400);
        res.json({ success: false, error: "Booking time not available" });
        return;
    }

    const [existingBooking] = await db
        .select()
        .from(bookingTable)
        .where(
            and(
                eq(bookingTable.time, selectedStart),
                eq(bookingTable.slotId, slot.id)
            )
        );

    if (existingBooking) {
        res.status(400);
        res.json({ success: false, error: "Booking time already taken" });
    }

    const [booking] = await db
        .insert(bookingTable)
        .values({
            patientId: req.body.patient_id,
            reason: req.body.reason,
            time: selectedStart,
            slotId: slot.id,
        })
        .returning();

    if (!booking) {
        res.status(400);
        res.json({ success: false, error: "Error while booking time" });

        return;
    }

    res.status(201);
    res.json({
        success: true,
        booking: {
            id: booking.id,
            patient_id: booking.patientId,
            reason: booking.reason,
            slot_id: booking.slotId,
            time: booking.time.toISOString(),
        },
    });
};

export const slotBookPost: RequestHandler[] = [
    validateRequest({
        body: slotBookPostBodySchema,
        params: slotBookPostParams,
    }),
    slotBookPostHandler as RequestHandler,
];
