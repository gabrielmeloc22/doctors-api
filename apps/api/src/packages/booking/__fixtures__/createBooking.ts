import { db } from "../../../database/db";
import { DeepPartial } from "../../../utils/deepPartial";
import { createSlot } from "../../slot/__fixtures__/createSlot";
import { bookingTable, IBooking } from "../bookingTable";

export const createBooking = async (
    args?: DeepPartial<IBooking>
): Promise<IBooking> => {
    const slotId = args?.slotId ? args.slotId : (await createSlot()).id;

    const [booking] = await db
        .insert(bookingTable)
        .values({
            slotId: slotId,
            time: new Date(),
            patientId: "1",
            reason: "checkup",
            ...args,
        })
        .returning();

    return booking as IBooking;
};
