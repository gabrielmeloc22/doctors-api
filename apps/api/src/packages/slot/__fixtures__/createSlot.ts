import { add } from "date-fns";
import { db } from "../../../database/db";
import { DeepPartial } from "../../../utils/deepPartial";
import { createDoctor } from "../../doctor/__fixtures__/createDoctor";
import { ISlot, SLOT_REPEAT_TYPE_ENUM, slotTable } from "../slotTable";

export const createSlot = async (args?: DeepPartial<ISlot>): Promise<ISlot> => {
    const doctor = await createDoctor();

    const [slot] = await db
        .insert(slotTable)
        .values({
            doctorId: doctor.id,
            duration: 30,
            startTime: new Date(),
            endTime: add(new Date(), { minutes: 30 }),
            repeatType: SLOT_REPEAT_TYPE_ENUM.SINGLE,
            ...args,
        })
        .returning();

    return slot as ISlot;
};
