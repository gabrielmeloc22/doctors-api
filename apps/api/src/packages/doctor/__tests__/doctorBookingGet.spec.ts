import request from "supertest";
import { expect, it } from "vitest";
import { createApp } from "../../../app";
import {
    defaultFrozenKeys,
    sanitizeTestObject,
} from "../../../testUtils/sanitizeTestObject";
import { createBooking } from "../../booking/__fixtures__/createBooking";
import { createSlot } from "../../slot/__fixtures__/createSlot";
import { SLOT_REPEAT_TYPE_ENUM } from "../../slot/slotTable";
import { createDoctor } from "../__fixtures__/createDoctor";

it("should validate params and return 400", async () => {
    const app = createApp();

    const response = await request(app.listen()).get(
        `/doctors/0a523035-fcca-4bbd-a7e6-3918e4689f56/bookings?start_time=12-21T10:00:00.000Z&end_time=24T10:00:00.000Z`
    );

    const body = response.body as Record<string, unknown>;

    expect(response.status).toBe(400);

    expect(body).toMatchSnapshot();
});

it("should get bookings within the date range", async () => {
    const app = createApp();

    const doctor = await createDoctor();
    const slot = await createSlot({
        duration: 30,
        doctorId: doctor.id,
        repeatType: SLOT_REPEAT_TYPE_ENUM.WEEKLY,
        startTime: new Date("2024-12-21T10:00:00.000Z"),
        endTime: new Date("2024-12-21T12:00:00.000Z"),
        repeatWeekdays: [1, 3, 5],
    });

    await createBooking({
        time: new Date("2024-12-23T10:00:00.000Z"),
        slotId: slot.id,
    });

    await createBooking({
        time: new Date("2024-12-23T10:30:00.000Z"),
        slotId: slot.id,
    });

    await createBooking({
        time: new Date("2024-12-25T10:00:00.000Z"),
        slotId: slot.id,
    });

    const response = await request(app.listen()).get(
        `/doctors/${doctor.id}/bookings?start_time=2024-12-21T10:00:00.000Z&end_time=2024-12-25T10:00:00.000Z`
    );

    const body = response.body as Record<string, unknown>;

    expect(response.status).toBe(200);

    expect(
        sanitizeTestObject({
            obj: body,
            frozenKeys: [...defaultFrozenKeys, "slot_id"],
        })
    ).toMatchSnapshot();
});
