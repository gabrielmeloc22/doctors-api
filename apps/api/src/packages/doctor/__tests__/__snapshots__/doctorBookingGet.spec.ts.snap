// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should get bookings within the date range 1`] = `
{
  "bookings": [
    {
      "id": "FROZEN-ID",
      "patient_id": "1",
      "reason": "checkup",
      "slot_id": "FROZEN-SLOT_ID",
      "time": "2024-12-23T10:00:00.000Z",
    },
    {
      "id": "FROZEN-ID",
      "patient_id": "1",
      "reason": "checkup",
      "slot_id": "FROZEN-SLOT_ID",
      "time": "2024-12-23T10:30:00.000Z",
    },
    {
      "id": "FROZEN-ID",
      "patient_id": "1",
      "reason": "checkup",
      "slot_id": "FROZEN-SLOT_ID",
      "time": "2024-12-25T10:00:00.000Z",
    },
  ],
  "success": true,
}
`;

exports[`should validate params and return 400 1`] = `
{
  "error": [
    {
      "message": "Invalid datetime",
      "path": [
        "start_time",
      ],
    },
    {
      "message": "Invalid datetime",
      "path": [
        "end_time",
      ],
    },
  ],
  "success": false,
}
`;
